rflx/parser/grammar.py

```diff
@@ -0,0 +1,574 @@
+    array_aggregate = (
+        Suppress(Literal("("))
+        + numeric_literal()
+        + (comma() - numeric_literal()) * (1,)
+        + Suppress(Literal(")"))
+    )
+    array_aggregate.setParseAction(parse_array_aggregate)
+
+    term = numeric_literal() | attribute_reference() | qualified_identifier() | array_aggregate
+    term.setParseAction(parse_term)
+
+    return (
+        infixNotation(
+            term, [(mathematical_operator, 2, opAssoc.LEFT, parse_mathematical_expression)]
+        )
+    ).setName("MathematicalExpression")
+
+
+def value_constraint() -> Token:
+    return (Keyword("if") - logical_expression()).setParseAction(lambda t: t[1])
+
+
+def type_derivation_definition() -> Token:
+    return (Keyword("new") - qualified_identifier()).setName("DerivationSpec")
+
+
+def size_aspect() -> Token:
+    return (Keyword("Size") - Keyword("=>") - mathematical_expression()).setParseAction(
+        parse_aspect
+    )
+
+
+def integer_type_definition() -> Token:
+    range_type_aspects = Keyword("with") - size_aspect()
+    range_type_aspects.setParseAction(parse_aspects)
+
+    range_type_definition = (
+        Keyword("range")
+        - mathematical_expression()
+        - Suppress(Literal(".."))
+        - mathematical_expression()
+        - range_type_aspects
+    )
+    range_type_definition.setName("RangeInteger")
+    modular_type_definition = Keyword("mod") - mathematical_expression()
+    modular_type_definition.setName("ModularInteger")
+
+    return range_type_definition | modular_type_definition
+
+
+def enumeration_type_definition() -> Token:
+    enumeration_literal = unqualified_identifier()
+    positional_enumeration = enumeration_literal + ZeroOrMore(comma() - enumeration_literal)
+    positional_enumeration.setParseAction(
+        lambda t: [(k, Number(v)) for v, k in enumerate(t.asList())]
+    )
+    element_value_association = enumeration_literal + Keyword("=>") - numeric_literal()
+    element_value_association.setParseAction(lambda t: (t[0], t[2]))
+    named_enumeration = element_value_association + ZeroOrMore(comma() - element_value_association)
+
+    boolean_literal = Keyword("True") | Keyword("False")
+    boolean_literal.setParseAction(lambda t: t[0] == "True")
+    boolean_aspect_definition = Optional(Keyword("=>") - boolean_literal)
+    boolean_aspect_definition.setParseAction(lambda t: (t if t else ["=>", True]))
+    always_valid_aspect = Literal("Always_Valid") - boolean_aspect_definition
+    always_valid_aspect.setParseAction(parse_aspect)
+    enumeration_aspects = Keyword("with") - delimitedList(size_aspect() | always_valid_aspect)
+    enumeration_aspects.setParseAction(parse_aspects)
+
+    return (
+        Literal("(")
+        - (named_enumeration | positional_enumeration)
+        - Literal(")")
+        - enumeration_aspects
+    ).setName("Enumeration")
+
+
+def array_type_definition() -> Token:
+    return (Keyword("array of") + unqualified_identifier()).setName("Array")
```

Comment:
This would not match "arrary  of" (multiple white space). `Keyword("array") + Keyword("of")` would be better.

Reply:
In an offline discussion we agreed that supporting anything but a single space is not desirable.